### ISSUE TITLE
Can the branches supporting CU early binding and Bridges resource config be merged into one branch?

### ISSUE THREAD

**Issue by mingtaiha** [radical-cybertools/radical.pilot#1369]
Title says it all.

I am referring specifically to branches `feature-early_binding` and `feature-bridges`

**Comment by andre-merzky** [comment]
I would actually be inclined to merge both into devel at this point. Can you please reconfirm that both, late binding and bridges support, work individually as expected?  Any objection to merging into devel?

**Comment by mingtaiha** [comment]
Do you mean early binding?

If so, then yes, the early binding features and Bridges config (except that MPI on Brigdes) both work. Any problems encountered along the way don't seem to be issues specific to changes made by the features.

**Comment by mingtaiha** [comment]
I get the following error when running on SuperMIC. It has to do with ORTE.

In the `agent_0.err`:
```
Traceback (most recent call last):
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/bin/radical-pilot-agent", line 26, in <module>
    rp.agent.bootstrap_3(agent_name)
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/bootstrap_3.py", line 42, in bootstrap_3
    agent = Agent_0(agent_name)
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/agent_0.py", line 117, in __init__
    session=self._session)
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/rm/base.py", line 220, in create
    return impl(cfg, session)
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/rm/torque.py", line 19, in __init__
    LRMS.__init__(self, cfg, session)
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/rm/base.py", line 150, in __init__
    rp.agent.LM.lrms_config_hook(lm, self._cfg, self, self._log))
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/lm/base.py", line 150, in lrms_config_hook
    from .orte           import ORTE
  File "/work/mha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017326.0008/pilot.0001/rp_install/lib/python2.7/site-packages/radical/pilot/agent/lm/orte.py", line 148
    logger.info('DVM stopped (%d)' % dvm_process.returncode)
         ^
SyntaxError: invalid syntax
```

In the `bootstrap_1.out`:
```
# -------------------------------------------------------------------
#
# install orte_cffi
# cmd: /worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/bin/pip install orte_cffi
#
Downloading/unpacking orte-cffi
  Downloading orte-cffi-0.4.1.tar.gz
  Running setup.py egg_info for package orte-cffi
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    c/_cffi_backend.c:15:17: error: ffi.h: No such file or directory
    In file included from c/_cffi_backend.c:63:
    c/malloc_closure.h:81: error: expected specifier-qualifier-list before ‘ffi_closure’
    c/malloc_closure.h: In function ‘more_core’:
    c/malloc_closure.h:117: warning: division by zero
    c/malloc_closure.h:149: error: ‘union mmaped_block’ has no member named ‘next’
    c/malloc_closure.h: At top level:
    c/malloc_closure.h:158: error: expected ‘)’ before ‘*’ token
    c/malloc_closure.h:166: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token
    c/_cffi_backend.c:259: error: expected specifier-qualifier-list before ‘ffi_cif’
    c/_cffi_backend.c: In function ‘cdataowninggc_dealloc’:
    c/_cffi_backend.c:1681: error: ‘ffi_closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1681: error: (Each undeclared identifier is reported only once
    c/_cffi_backend.c:1681: error: for each function it appears in.)
    c/_cffi_backend.c:1681: error: ‘closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1681: error: expected expression before ‘)’ token
    c/_cffi_backend.c:1684: warning: implicit declaration of function ‘cffi_closure_free’
    c/_cffi_backend.c: In function ‘cdataowninggc_traverse’:
    c/_cffi_backend.c:1701: error: ‘ffi_closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1701: error: ‘closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1701: error: expected expression before ‘)’ token
    c/_cffi_backend.c: In function ‘cdataowninggc_clear’:
    c/_cffi_backend.c:1722: error: ‘ffi_closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1722: error: ‘closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1722: error: expected expression before ‘)’ token
    c/_cffi_backend.c: In function ‘cdataowninggc_repr’:
    c/_cffi_backend.c:1928: error: ‘ffi_closure’ undeclared (first use in this function)
    c/_cffi_backend.c:1928: error: expected expression before ‘)’ token
    c/_cffi_backend.c: At top level:
    c/_cffi_backend.c:2618: error: expected declaration specifiers or ‘...’ before ‘ffi_abi’
    c/_cffi_backend.c: In function ‘cdata_call’:
    c/_cffi_backend.c:2751: error: ‘ffi_abi’ undeclared (first use in this function)
    c/_cffi_backend.c:2751: error: expected ‘;’ before ‘fabi’
    c/_cffi_backend.c:2793: error: ‘fabi’ undeclared (first use in this function)
    c/_cffi_backend.c:2797: error: too many arguments to function ‘fb_prepare_cif’
    c/_cffi_backend.c:2802: error: ‘cif_description_t’ has no member named ‘exchange_size’
    c/_cffi_backend.c:2812: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’
    c/_cffi_backend.c:2842: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’
    c/_cffi_backend.c:2847: warning: implicit declaration of function ‘ffi_call’
    c/_cffi_backend.c:2847: error: ‘cif_description_t’ has no member named ‘cif’
    c/_cffi_backend.c: In function ‘new_primitive_type’:
    c/_cffi_backend.c:4042: error: ‘ffi_type’ undeclared (first use in this function)
    c/_cffi_backend.c:4042: error: ‘ffitype’ undeclared (first use in this function)
    c/_cffi_backend.c:4060: error: ‘ffi_type_sint8’ undeclared (first use in this function)
    c/_cffi_backend.c:4061: error: ‘ffi_type_sint16’ undeclared (first use in this function)
    c/_cffi_backend.c:4062: error: ‘ffi_type_sint32’ undeclared (first use in this function)
    c/_cffi_backend.c:4063: error: ‘ffi_type_sint64’ undeclared (first use in this function)
    c/_cffi_backend.c:4069: error: ‘ffi_type_float’ undeclared (first use in this function)
    c/_cffi_backend.c:4071: error: ‘ffi_type_double’ undeclared (first use in this function)
    c/_cffi_backend.c:4079: error: ‘ffi_type_longdouble’ undeclared (first use in this function)
    c/_cffi_backend.c:4086: error: ‘ffi_type_uint8’ undeclared (first use in this function)
    c/_cffi_backend.c:4087: error: ‘ffi_type_uint16’ undeclared (first use in this function)
    c/_cffi_backend.c:4088: error: ‘ffi_type_uint32’ undeclared (first use in this function)
    c/_cffi_backend.c:4089: error: ‘ffi_type_uint64’ undeclared (first use in this function)
    c/_cffi_backend.c: At top level:
    c/_cffi_backend.c:4736: error: expected specifier-qualifier-list before ‘ffi_type’
    c/_cffi_backend.c:4761: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token
    c/_cffi_backend.c:4771: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token
    c/_cffi_backend.c: In function ‘fb_build’:
    c/_cffi_backend.c:4923: error: ‘struct funcbuilder_s’ has no member named ‘atypes’
    c/_cffi_backend.c:4923: error: ‘ffi_type’ undeclared (first use in this function)
    c/_cffi_backend.c:4923: error: expected expression before ‘)’ token
    c/_cffi_backend.c:4924: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:4927: error: ‘struct funcbuilder_s’ has no member named ‘rtype’
    c/_cffi_backend.c:4927: warning: implicit declaration of function ‘fb_fill_type’
    c/_cffi_backend.c:4935: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’
    c/_cffi_backend.c:4938: error: ‘struct funcbuilder_s’ has no member named ‘rtype’
    c/_cffi_backend.c:4939: error: ‘ffi_arg’ undeclared (first use in this function)
    c/_cffi_backend.c:4949: error: ‘atype’ undeclared (first use in this function)
    c/_cffi_backend.c:4962: error: ‘struct funcbuilder_s’ has no member named ‘atypes’
    c/_cffi_backend.c:4963: error: ‘struct funcbuilder_s’ has no member named ‘atypes’
    c/_cffi_backend.c:4966: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’
    c/_cffi_backend.c:4975: error: ‘cif_description_t’ has no member named ‘exchange_size’
    c/_cffi_backend.c: In function ‘fb_build_name’:
    c/_cffi_backend.c:4999: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5011: error: ‘struct funcbuilder_s’ has no member named ‘fct’
    c/_cffi_backend.c:5014: error: ‘struct funcbuilder_s’ has no member named ‘fct’
    c/_cffi_backend.c: In function ‘fb_prepare_ctype’:
    c/_cffi_backend.c:5058: error: ‘struct funcbuilder_s’ has no member named ‘fct’
    c/_cffi_backend.c:5075: error: ‘struct funcbuilder_s’ has no member named ‘fct’
    c/_cffi_backend.c: At top level:
    c/_cffi_backend.c:5095: error: expected declaration specifiers or ‘...’ before ‘ffi_abi’
    c/_cffi_backend.c: In function ‘fb_prepare_cif’:
    c/_cffi_backend.c:5122: warning: implicit declaration of function ‘ffi_prep_cif’
    c/_cffi_backend.c:5122: error: ‘cif_description_t’ has no member named ‘cif’
    c/_cffi_backend.c:5122: error: ‘fabi’ undeclared (first use in this function)
    c/_cffi_backend.c:5122: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5123: error: ‘struct funcbuilder_s’ has no member named ‘rtype’
    c/_cffi_backend.c:5123: error: ‘struct funcbuilder_s’ has no member named ‘atypes’
    c/_cffi_backend.c:5123: error: ‘FFI_OK’ undeclared (first use in this function)
    c/_cffi_backend.c: In function ‘new_function_type’:
    c/_cffi_backend.c:5167: error: too many arguments to function ‘fb_prepare_cif’
    c/_cffi_backend.c:5181: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5191: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5201: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5204: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5205: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c:5207: error: ‘struct funcbuilder_s’ has no member named ‘nargs’
    c/_cffi_backend.c: In function ‘b_new_function_type’:
    c/_cffi_backend.c:5218: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)
    c/_cffi_backend.c: In function ‘convert_from_object_fficallback’:
    c/_cffi_backend.c:5238: error: ‘ffi_arg’ undeclared (first use in this function)
    c/_cffi_backend.c: At top level:
    c/_cffi_backend.c:5420: error: expected ‘)’ before ‘*’ token
    c/_cffi_backend.c: In function ‘prepare_callback_info_tuple’:
    c/_cffi_backend.c:5462: error: ‘ffi_arg’ undeclared (first use in this function)
    c/_cffi_backend.c: In function ‘b_callback’:
    c/_cffi_backend.c:5495: error: ‘ffi_closure’ undeclared (first use in this function)
    c/_cffi_backend.c:5495: error: ‘closure’ undeclared (first use in this function)
    c/_cffi_backend.c:5505: warning: implicit declaration of function ‘cffi_closure_alloc’
    c/_cffi_backend.c:5523: warning: implicit declaration of function ‘ffi_prep_closure’
    c/_cffi_backend.c:5523: error: ‘cif_description_t’ has no member named ‘cif’
    c/_cffi_backend.c:5524: error: ‘invoke_callback’ undeclared (first use in this function)
    c/_cffi_backend.c:5524: error: ‘FFI_OK’ undeclared (first use in this function)
    In file included from c/cffi1_module.c:3,
                     from c/_cffi_backend.c:6896:
    c/realize_c_type.c: In function ‘realize_c_type_or_func’:
    c/realize_c_type.c:577: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)
    c/_cffi_backend.c: In function ‘init_cffi_backend’:
    c/_cffi_backend.c:6988: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)
    Traceback (most recent call last):
      File "<string>", line 16, in <module>
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/build/orte-cffi/setup.py", line 15, in <module>
        install_requires = ["cffi>=1.5.0"],
      File "/usr/local/packages/python/2.7.7-anaconda/lib/python2.7/distutils/core.py", line 111, in setup
        _setup_distribution = dist = klass(attrs)
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 260, in __init__
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 284, in fetch_build_eggs
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 563, in resolve
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 799, in best_match
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 811, in obtain
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 327, in fetch_build_egg
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 446, in easy_install
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 476, in install_item
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 655, in install_eggs
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 930, in build_and_install
      File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 921, in run_setup
    distutils.errors.DistutilsError: Setup script exited with error: command 'gcc' failed with exit status 1
    Complete output from command python setup.py egg_info:
    Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

c/_cffi_backend.c:15:17: error: ffi.h: No such file or directory

In file included from c/_cffi_backend.c:63:

c/malloc_closure.h:81: error: expected specifier-qualifier-list before ‘ffi_closure’

c/malloc_closure.h: In function ‘more_core’:

c/malloc_closure.h:117: warning: division by zero

c/malloc_closure.h:149: error: ‘union mmaped_block’ has no member named ‘next’

c/malloc_closure.h: At top level:

c/malloc_closure.h:158: error: expected ‘)’ before ‘*’ token

c/malloc_closure.h:166: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token

c/_cffi_backend.c:259: error: expected specifier-qualifier-list before ‘ffi_cif’

c/_cffi_backend.c: In function ‘cdataowninggc_dealloc’:

c/_cffi_backend.c:1681: error: ‘ffi_closure’ undeclared (first use in this function)

c/_cffi_backend.c:1681: error: (Each undeclared identifier is reported only once

c/_cffi_backend.c:1681: error: for each function it appears in.)

c/_cffi_backend.c:1681: error: ‘closure’ undeclared (first use in this function)

c/_cffi_backend.c:1681: error: expected expression before ‘)’ token

c/_cffi_backend.c:1684: warning: implicit declaration of function ‘cffi_closure_free’

c/_cffi_backend.c: In function ‘cdataowninggc_traverse’:

c/_cffi_backend.c:1701: error: ‘ffi_closure’ undeclared (first use in this function)

c/_cffi_backend.c:1701: error: ‘closure’ undeclared (first use in this function)

c/_cffi_backend.c:1701: error: expected expression before ‘)’ token

c/_cffi_backend.c: In function ‘cdataowninggc_clear’:

c/_cffi_backend.c:1722: error: ‘ffi_closure’ undeclared (first use in this function)

c/_cffi_backend.c:1722: error: ‘closure’ undeclared (first use in this function)

c/_cffi_backend.c:1722: error: expected expression before ‘)’ token

c/_cffi_backend.c: In function ‘cdataowninggc_repr’:

c/_cffi_backend.c:1928: error: ‘ffi_closure’ undeclared (first use in this function)

c/_cffi_backend.c:1928: error: expected expression before ‘)’ token

c/_cffi_backend.c: At top level:

c/_cffi_backend.c:2618: error: expected declaration specifiers or ‘...’ before ‘ffi_abi’

c/_cffi_backend.c: In function ‘cdata_call’:

c/_cffi_backend.c:2751: error: ‘ffi_abi’ undeclared (first use in this function)

c/_cffi_backend.c:2751: error: expected ‘;’ before ‘fabi’

c/_cffi_backend.c:2793: error: ‘fabi’ undeclared (first use in this function)

c/_cffi_backend.c:2797: error: too many arguments to function ‘fb_prepare_cif’

c/_cffi_backend.c:2802: error: ‘cif_description_t’ has no member named ‘exchange_size’

c/_cffi_backend.c:2812: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’

c/_cffi_backend.c:2842: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’

c/_cffi_backend.c:2847: warning: implicit declaration of function ‘ffi_call’

c/_cffi_backend.c:2847: error: ‘cif_description_t’ has no member named ‘cif’

c/_cffi_backend.c: In function ‘new_primitive_type’:

c/_cffi_backend.c:4042: error: ‘ffi_type’ undeclared (first use in this function)

c/_cffi_backend.c:4042: error: ‘ffitype’ undeclared (first use in this function)

c/_cffi_backend.c:4060: error: ‘ffi_type_sint8’ undeclared (first use in this function)

c/_cffi_backend.c:4061: error: ‘ffi_type_sint16’ undeclared (first use in this function)

c/_cffi_backend.c:4062: error: ‘ffi_type_sint32’ undeclared (first use in this function)

c/_cffi_backend.c:4063: error: ‘ffi_type_sint64’ undeclared (first use in this function)

c/_cffi_backend.c:4069: error: ‘ffi_type_float’ undeclared (first use in this function)

c/_cffi_backend.c:4071: error: ‘ffi_type_double’ undeclared (first use in this function)

c/_cffi_backend.c:4079: error: ‘ffi_type_longdouble’ undeclared (first use in this function)

c/_cffi_backend.c:4086: error: ‘ffi_type_uint8’ undeclared (first use in this function)

c/_cffi_backend.c:4087: error: ‘ffi_type_uint16’ undeclared (first use in this function)

c/_cffi_backend.c:4088: error: ‘ffi_type_uint32’ undeclared (first use in this function)

c/_cffi_backend.c:4089: error: ‘ffi_type_uint64’ undeclared (first use in this function)

c/_cffi_backend.c: At top level:

c/_cffi_backend.c:4736: error: expected specifier-qualifier-list before ‘ffi_type’

c/_cffi_backend.c:4761: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token

c/_cffi_backend.c:4771: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘*’ token

c/_cffi_backend.c: In function ‘fb_build’:

c/_cffi_backend.c:4923: error: ‘struct funcbuilder_s’ has no member named ‘atypes’

c/_cffi_backend.c:4923: error: ‘ffi_type’ undeclared (first use in this function)

c/_cffi_backend.c:4923: error: expected expression before ‘)’ token

c/_cffi_backend.c:4924: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:4927: error: ‘struct funcbuilder_s’ has no member named ‘rtype’

c/_cffi_backend.c:4927: warning: implicit declaration of function ‘fb_fill_type’

c/_cffi_backend.c:4935: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’

c/_cffi_backend.c:4938: error: ‘struct funcbuilder_s’ has no member named ‘rtype’

c/_cffi_backend.c:4939: error: ‘ffi_arg’ undeclared (first use in this function)

c/_cffi_backend.c:4949: error: ‘atype’ undeclared (first use in this function)

c/_cffi_backend.c:4962: error: ‘struct funcbuilder_s’ has no member named ‘atypes’

c/_cffi_backend.c:4963: error: ‘struct funcbuilder_s’ has no member named ‘atypes’

c/_cffi_backend.c:4966: error: ‘cif_description_t’ has no member named ‘exchange_offset_arg’

c/_cffi_backend.c:4975: error: ‘cif_description_t’ has no member named ‘exchange_size’

c/_cffi_backend.c: In function ‘fb_build_name’:

c/_cffi_backend.c:4999: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5011: error: ‘struct funcbuilder_s’ has no member named ‘fct’

c/_cffi_backend.c:5014: error: ‘struct funcbuilder_s’ has no member named ‘fct’

c/_cffi_backend.c: In function ‘fb_prepare_ctype’:

c/_cffi_backend.c:5058: error: ‘struct funcbuilder_s’ has no member named ‘fct’

c/_cffi_backend.c:5075: error: ‘struct funcbuilder_s’ has no member named ‘fct’

c/_cffi_backend.c: At top level:

c/_cffi_backend.c:5095: error: expected declaration specifiers or ‘...’ before ‘ffi_abi’

c/_cffi_backend.c: In function ‘fb_prepare_cif’:

c/_cffi_backend.c:5122: warning: implicit declaration of function ‘ffi_prep_cif’

c/_cffi_backend.c:5122: error: ‘cif_description_t’ has no member named ‘cif’

c/_cffi_backend.c:5122: error: ‘fabi’ undeclared (first use in this function)

c/_cffi_backend.c:5122: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5123: error: ‘struct funcbuilder_s’ has no member named ‘rtype’

c/_cffi_backend.c:5123: error: ‘struct funcbuilder_s’ has no member named ‘atypes’

c/_cffi_backend.c:5123: error: ‘FFI_OK’ undeclared (first use in this function)

c/_cffi_backend.c: In function ‘new_function_type’:

c/_cffi_backend.c:5167: error: too many arguments to function ‘fb_prepare_cif’

c/_cffi_backend.c:5181: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5191: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5201: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5204: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5205: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c:5207: error: ‘struct funcbuilder_s’ has no member named ‘nargs’

c/_cffi_backend.c: In function ‘b_new_function_type’:

c/_cffi_backend.c:5218: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)

c/_cffi_backend.c: In function ‘convert_from_object_fficallback’:

c/_cffi_backend.c:5238: error: ‘ffi_arg’ undeclared (first use in this function)

c/_cffi_backend.c: At top level:

c/_cffi_backend.c:5420: error: expected ‘)’ before ‘*’ token

c/_cffi_backend.c: In function ‘prepare_callback_info_tuple’:

c/_cffi_backend.c:5462: error: ‘ffi_arg’ undeclared (first use in this function)

c/_cffi_backend.c: In function ‘b_callback’:

c/_cffi_backend.c:5495: error: ‘ffi_closure’ undeclared (first use in this function)

c/_cffi_backend.c:5495: error: ‘closure’ undeclared (first use in this function)

c/_cffi_backend.c:5505: warning: implicit declaration of function ‘cffi_closure_alloc’

c/_cffi_backend.c:5523: warning: implicit declaration of function ‘ffi_prep_closure’

c/_cffi_backend.c:5523: error: ‘cif_description_t’ has no member named ‘cif’

c/_cffi_backend.c:5524: error: ‘invoke_callback’ undeclared (first use in this function)

c/_cffi_backend.c:5524: error: ‘FFI_OK’ undeclared (first use in this function)

In file included from c/cffi1_module.c:3,

                 from c/_cffi_backend.c:6896:

c/realize_c_type.c: In function ‘realize_c_type_or_func’:

c/realize_c_type.c:577: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)

c/_cffi_backend.c: In function ‘init_cffi_backend’:

c/_cffi_backend.c:6988: error: ‘FFI_DEFAULT_ABI’ undeclared (first use in this function)

Traceback (most recent call last):

  File "<string>", line 16, in <module>

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/build/orte-cffi/setup.py", line 15, in <module>

    install_requires = ["cffi>=1.5.0"],

  File "/usr/local/packages/python/2.7.7-anaconda/lib/python2.7/distutils/core.py", line 111, in setup

    _setup_distribution = dist = klass(attrs)

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 260, in __init__

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 284, in fetch_build_eggs

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 563, in resolve

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 799, in best_match

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 811, in obtain

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 327, in fetch_build_egg

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 446, in easy_install

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 476, in install_item

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 655, in install_eggs

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 930, in build_and_install

  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 921, in run_setup

distutils.errors.DistutilsError: Setup script exited with error: command 'gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command python setup.py egg_info failed with error code 1 in /worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/build/orte-cffi
Traceback (most recent call last):
  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/bin/pip", line 8, in <module>
    load_entry_point('pip==1.4.1', 'console_scripts', 'pip')()
  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/pip/__init__.py", line 148, in main
    return command.main(args[1:], options)
  File "/worka/work/mha/radical.pilot.sandbox/ve.xsede.supermic.0.46/lib/python2.7/site-packages/pip/basecommand.py", line 169, in main
    text = '\n'.join(complete_log)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 75: ordinal not in range(128)
#
# ERROR
# no fallback command available
#
# -------------------------------------------------------------------
Couldn't install orte_cffi! Lets see how far we get ...
```

**Comment by andre-merzky** [comment]
Thanks for testing, Ming.  Two issues at play here - one in the code, which I fixed, and one in ORTE deployment on supermic, which I have to look into :/

**Comment by andre-merzky** [comment]
wait - devel does not have ORTE-support for supermic, so the error in the bootstrapper does not matter (the bootstrapper is not clever enough to know that orte-cffi is not needed).  Could you please give devel yet another try?

**Comment by mingtaiha** [comment]
I am able to submit jobs to SuperMIC (with some failing), but now I get the ORTE error on Comet.

```
Traceback (most recent call last):
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/utils/component.py", line 528, in ru_initialize_common
    self._log)
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/utils/component.py", line 251, in start_components
    comp = ctype.create(tmp_cfg, session)
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/agent/executing/base.py", line 66, in create
    from .orte import ORTE
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/agent/executing/orte.py", line 15, in <module>
    from orte_cffi import ffi, lib as orte_lib
ImportError: No module named orte_cffi
2017-06-12 14:51:25,251: agent_0             : MainProcess                     : MainThread     : ERROR   : initialization error
Traceback (most recent call last):
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/utils/process.py", line 776, in _ru_initialize
    self.ru_initialize_common()
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/utils/component.py", line 528, in ru_initialize_common
    self._log)
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/utils/component.py", line 251, in start_components
    comp = ctype.create(tmp_cfg, session)
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/agent/executing/base.py", line 66, in create
    from .orte import ORTE
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/pilot/agent/executing/orte.py", line 15, in <module>
    from orte_cffi import ffi, lib as orte_lib
ImportError: No module named orte_cffi
2017-06-12 14:51:25,252: agent_0             : MainProcess                     : MainThread     : ERROR   : initialization failed (initialize: ImportError('No module named orte_cffi',))
Traceback (most recent call last):
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/utils/process.py", line 468, in start
    self._ru_initialize()
  File "/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017329.0004/pilot.0000/rp_install/lib/python2.7/site-packages/radical/utils/process.py", line 793, in _ru_initialize
    raise RuntimeError('initialize: %s' % repr(e))
RuntimeError: initialize: ImportError('No module named orte_cffi',)
```

My current `radical-stack`:
```
  python               : 2.7.12
  pythonpath           : 
  virtualenv           : /home/mingtha/ve/ve.exec_model_exp

  radical.pilot        : 0.46-split-84-g1219765a@devel
  radical.utils        : 0.46-v0.45-49-g38f6ee3@devel
  saga                 : 0.46-split-19-g11be22af@devel
```

**Comment by andre-merzky** [comment]
If this persists after the release, please do open a new ticket for the comet problem.  The original point (the merge) should be addressed, so I am going to close this ticket.
